quinn-proto/src/token.rs

```diff
@@ -43,46 +49,55 @@ impl RetryToken {
                 .unwrap_or(0),
         );
 
-        let signature_pos = buf.len();
+        let mut additional_data = Vec::new();
         match address.ip() {
-            IpAddr::V4(x) => buf.put_slice(&x.octets()),
-            IpAddr::V6(x) => buf.put_slice(&x.octets()),
+            IpAddr::V4(x) => additional_data.put_slice(&x.octets()),
+            IpAddr::V6(x) => additional_data.put_slice(&x.octets()),
         }
-        buf.write(address.port());
-        retry_src_cid.encode_long(&mut buf);
-
-        let signature = key.sign(&buf);
-        // No reason to actually encode the IP in the token, since we always have the remote addr for an incoming packet.
-        buf.truncate(signature_pos);
-        buf.extend_from_slice(signature.as_ref());
-        buf
+        additional_data.write(address.port());
+        retry_src_cid.encode_long(&mut additional_data);
+
+        aead_key.seal(&mut buf, &additional_data).unwrap();
+
+        let mut token = Vec::new();
+        token.put_slice(self.random_bytes);
+        token.put_slice(&buf);
+        token
     }
 
     pub fn from_bytes(
-        key: &impl HmacKey,
+        key: &impl PseudoRandomKey,
         address: &SocketAddr,
         retry_src_cid: &ConnectionId,
-        data: &[u8],
+        raw_token_bytes: &'a [u8],
     ) -> Result<Self, ()> {
-        let mut reader = io::Cursor::new(data);
+        if raw_token_bytes.len() < Self::RANDOM_BYTES_LEN {
+            // Invalid length
+            return Err(());
+        }
 
-        let orig_dst_cid = ConnectionId::decode_long(&mut reader).ok_or(())?;
-        let issued = UNIX_EPOCH + Duration::new(reader.get::<u64>().map_err(|_| ())?, 0);
+        let random_bytes = &raw_token_bytes[..Self::RANDOM_BYTES_LEN];
+        let aead_key: ring::aead::LessSafeKey = AeadKey::from_hkdf(key, random_bytes);
+        let mut sealed_token = raw_token_bytes[Self::RANDOM_BYTES_LEN..].to_vec();
 
-        let signature_start = reader.position() as usize;
-        let mut buf = Vec::new();
-        buf.put_slice(&data[0..signature_start]);
+        let mut additional_data = Vec::new();
```

Comment:
Since it looks like the maximum length of the additional data is known in advance and is pretty limited, it'd be nice to avoid the heap allocation here and use an array instead.

Reply:
Use the `BufMut` impl for `&mut [u8]`:
```rust
let mut additional_data = [0u8; 16 + 2]; // max(ipv4, ipv6) + port
let mut cursor = &mut additional_data[..];
cursor.put_slice(etc);
```

Reply:
Also take care to only feed in the portion of `additional_data` that was actually written, otherwise we might have a collision issue.

Reply:
@Ralith This was really helpful, thanks :+1: